### DISCUSSION TITLE
Serial communication issue with ATTINY816

### DISCUSSION THREAD

**Discussion by delucaide** [SpenceKonde/megaTinyCore#789] [Q&A]
Hello everyone,
i'm getting problem when i try to estabilish a serial communication with the attiny816.
Currently i get non correct characters output. Anyone of you had some issue like this? 
i checked the baudrate with an oscilloscope while printing "a" and seems correct.

![oscilloscope a](https://user-images.githubusercontent.com/33809522/188174710-d53b70fb-48c6-4154-a746-6da6756170b5.PNG)
![Printing 0](https://user-images.githubusercontent.com/33809522/188174717-6342788f-4ac2-4778-9e8d-5ad1c3f64383.PNG)
![Printing a](https://user-images.githubusercontent.com/33809522/188174726-c152ad92-6c75-49d1-ab88-8eb5aa98f604.PNG)
![Printing b](https://user-images.githubusercontent.com/33809522/188174728-73439bd0-60ec-4ea2-9332-c124b64fa6ad.PNG)
![Settings](https://user-images.githubusercontent.com/33809522/188174735-8ceedd8e-f5ed-48d1-9520-25c85acea5e7.png)

**Comment by mosqu-ito** [comment]
Have you tried it at the default chip speed of 20 MHz?

**Reply by SpenceKonde** [reply]
shouldn't matter - we set the fuses for the proper clock speed when uploading via UPDI

**Reply by SpenceKonde** [reply]
That's not the github version right? Don't use the github version right now - it is not in a usable state atm I don't think and I'm planning to work on that asap but have to finish work for a client.

**Comment by SpenceKonde** [comment]
Hm Unless my eyes are playing tricks on me, that;'s no A you're seeing on
the scope; I wager that where it's showing a printing character, what you
see on the scope is not what your sketch is trying to send....

Which implies that you are NOT using the most recent released version,
because I have used serial on that dozens of times and it's fine, whereas
it's one of the things that may have been broken. It's getting the
baud right but sending wrong char (or rather, something is wrong with my ,
but is sending garbage. And as I said, major parts of mTC in github version
are thoroughly hosed. Because I keep gettign pulled off of this for some
urgent thing or another

On Fri, Sep 2, 2022 at 10:47 AM delucaide ***@***.***> wrote:

> Hello everyone,
> i'm getting problem when i try to estabilish a serial communication with
> the attiny816.
> Currently i get non correct characters output. Anyone of you had some
> issue like this?
> i checked the baudrate with an oscilloscope while printing "a" and seems
> correct.
>
> [image: oscilloscope a]
> <https://user-images.githubusercontent.com/33809522/188174710-d53b70fb-48c6-4154-a746-6da6756170b5.PNG>
> [image: Printing 0]
> <https://user-images.githubusercontent.com/33809522/188174717-6342788f-4ac2-4778-9e8d-5ad1c3f64383.PNG>
> [image: Printing a]
> <https://user-images.githubusercontent.com/33809522/188174726-c152ad92-6c75-49d1-ab88-8eb5aa98f604.PNG>
> [image: Printing b]
> <https://user-images.githubusercontent.com/33809522/188174728-73439bd0-60ec-4ea2-9332-c124b64fa6ad.PNG>
> [image: Settings]
> <https://user-images.githubusercontent.com/33809522/188174735-8ceedd8e-f5ed-48d1-9520-25c85acea5e7.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/789>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW3SC3B2NLRMXWM6IXDV4IHOVANCNFSM6AAAAAAQDK7E6Q>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by SpenceKonde** [reply]
Okay, he's not using latest from github, otherwise he'd have other complaints ("nothing compiles")

**Reply by delucaide** [reply]
Hello,
i'm using the 2.5.11 version downloaded from the board manager. Can u suggest any test i can do?

**Reply by delucaide** [reply]
i always get the " , " instead of "a" when i serial print and both sender and receiver baudrate match. i tried also to downgrade the bootloader without any changes

**Reply by delucaide** [reply]
Sorry, my mistake. I get
 "0" when i print "a"
" ' " when i print "b"
" N " when i print "c"

same result with software serial also

**Reply by mosqu-ito** [reply]
It was smart trying software serial, and rather amazing that the problem persisted then.  I guess there's not much point trying it without the swap (PB2) then except perhaps to ensure that it's not something to do with pin PA1 itself).  The repeatability suggests it's not noise or a bad connection, but it doesn't seem to point to any obvious translation that would result from a framing error either.  All I can think to ask you to do is to print much slower (1000mS) to confirm that it is at least a 1:1 translation and not a 1:2 or a 2:1.

**Comment by SpenceKonde** [comment]
Wait a minute here.... You were seeing that scope trace when trying to print 'a'
No, my eyes *were* deceiving me. **That's undoubtedly a correct 'a' on the scope trace, clear as crystal.** 

Serial bits are sent in reverse order from every other protocol for some reason (which I think is most likely both moronic and historical in nature - serial and the ascii encoding are far older than SPI and I2C), and the first bit sent is always a zero (the start bit), preceeding the data bits, then the LSBit is sent first, followed by the other bits in ascending order, after the MSBit of data, the line is left high for a minimum of one baud period (This is the stop bit; a framing error gets recorded if where it expects the stop bit to be isn't HIGH). That's only the minimum length of the high period before the next character (when sending multiple characters, assuming the baud rate is not so high that the chip has trouble delivering data to the USART that fast, as soon as this stop bit has ended, 10 baud periods after the beginning of the first start bit, you would expect to see the start bit of the second character begin). Of course, if it has nothing more to send, the line will just stay high. It can be set to use two stop bits, which is typically done when talking to something that may struggle otherwise due to for example differences in clock speed, or a struggling software serial implementation that's not able to get ready to receive another character during the 1.5 bit periods between sampling the last data bit and when it would have to react to the next start bit. 

So... that scope traces out 0100001101    which means 10000110 after dropping start and stop, and 01100001 in the normal order or 0x61. **A correctly formed 'a'** 

Those values you are seeing look suspiciously like a bad attempt to interpret data assuming the MSB first bit order

Since the scope agrees that the tiny is sending the correct signal, and I trust a scope more than I trust an unknown receiver.
Another thing there that raises my eyebrows is the amount of noise on the highs - that looks like about a half volt high frequency noise, as well as the apparent 0.7V difference between the minimum voltage and ground, if I am reading those measurements at the bottom correctly. Can you confirm or explain this? 

Your receiver and/or terminal program appears to be malfunctioning. Might the ground connection between the two not be making contact (or be absent altogether)? Might it be expecting the 

Something is up with power too. If the micro really is only able to drive the line to 0.7V, that also indicates that something is wrong - I would expect to see it going almost all the way to ground - only the required sink current to overcome whatever is pulling up the line, and in this case that implies that the poor micro is sinking 28mA to do so! (see fig 35-38 in the datasheet), and extrapolate off the righthand edge of the graph. It's also higher than the recommended maximum per-pin current by 40%, though still  well clear of the 40 mA abs. maximum.  That should not be the case, the current involved in serial should be tiny. Poor connection to ground could result in the ground  at the microcontroller drifting up with less current than that, as the current it's sinking drags the ground rail upwards. And you shouldn't see that noise (half a volt) on the TX line. 

Some reality checks are in order - confirm what the difference between Vcc and Gnd is on the micro (measure, ideally, the micro's pins themself first with the DVM, then put a scope between those points and and see whether the same noise is seen there as you're seeing on the HIGHs it outputs. If so you should fix your power supply circuitry and ensure that appropriate decoupling is present - both chip-level (0.1uf ceramic cap right next to the chip) and board-level (I use a 4.7 uF ceramic cap not smaller in physical size than 0805 for 5v applications, unless requirements of the application dictate otherwise. Though I absolutely don't think this why the serial being read wrong, a problem with the power supply circuit can definitely cause other problems. The main reason I am suggesting this reality check is because of how far from ground the lows appear to be - the chip seems to have to work unreasonably hard to drive the pin low, which should not be the case. The investigation of these issues may lead to insights into the serial issue, or it may not. But it will be necessary sooner or later in any event, 

(0603 is too small for 4.7uF @ 5V with current process technology, most _but not all_ 4.7uF nom. capacitors in 0805 are good for 5v. This spec isn't in the datasheets, have to get it from mfg. website (murata, tdk, samsung all have pretty decent tools) will tell you and often give you a graph. Charge and energy are physical quantities, and they cannot be stored in arbitrarily small volumes; on ceramic caps this manifests as a significant dependence of capacitance on DC bias. On electrolytic or tantalum caps, that manifests as a limit on the maximum capacitance and voltage for any given size. Not that that isn't true of ceramics as well, but you routinely see things with crazy specs, like 0805 ceramics rated for 22uF and 35V, which sounds too good to be true. Because it is and the C vs DC bias curve on those caps shows that by 5V, they're giving you under 10uF, plummeting further to just 1.5uF at 25V, and 1uF at 35V)

**Reply by mosqu-ito** [reply]
I too noticed the voltages, the relative Vpp looking very good but the absolute voltages being strangely high.  And yet both ones and zeros are arriving; the problem, whatever it is, scrambles the data in a consistent and repeatable fashion, albeit an as yet indecipherable one.  How about sending exactly 256 bytes 0 through 255 and showing us what arrives then?  It should be interesting if nothing else.

**Reply by delucaide** [reply]
Can u explain how to do this test?

**Reply by mosqu-ito** [reply]
<pre>
void setup() {

   word c;

   delay(5000);  // In case you need a little time to get ready to receive.
   delay(1000);
   Serial.begin(9600);
   delay(1000);

   Serial.println();
   Serial.println("Here we go...");
   Serial.println();

   // for (c=0; c<=255; c++) {   // Optionally skip control character range so it looks tidy.
   for (c=21; c<=255; c++) {
      Serial.print("[");
      Serial.printf("%3d",c);    // Lets us specify the width, again just to be prettier.
      Serial.print("]=["); Serial.print(char(c)); Serial.print("] ");
      if ((c % 10) == 0) Serial.println();  // Let's not make it all one line.
      delay(50);  // Making REAL sure we don't overrun the buffer.
      }

   Serial.println();
   Serial.println();
   Serial.println("...all done.");
   Serial.println();

   delay(1000);

   }


void loop() {}
</pre>

I don't use Windows but I took a peek at your RealTerm and to call it feature-rich would be an understatement.  It certainly gives a fellow plenty of rope with which to hang himself.  I notice there's a whole tab about encoding.  Maybe we _are_ looking at a 2:1 translation?

**Reply by mosqu-ito** [reply]
All that formatting looks oh-so "pretty" on _my working_ comm connection.    :D 
But of course on your connection it's just going to be more garbage to sift through,
even the line-feeds.  You'd better start with a stripped down version like so...

<pre>
void setup() {

   word c;

   delay(5000);          // In case you need a little time to get ready to receive.
   Serial.begin(9600);
   delay(1000);

   for (c=0; c<=255; c++) {
      Serial.print(char(c));
      delay(50);
      }

   delay(1000);

   }


void loop() {}

</pre>

**Reply by SpenceKonde** [reply]
For loop syntax is a bit off. I'd do this construction. That's how you get 256 iterations with a 1 byte counter variable - the do....while construct always executes the body, so while c=0 doesn't satisfy the while condition, by the time of first test c = 1. then it keeps going until the loop that starts with c=255, writes that, postincrement of c rolls it over, and then it fails the test in the while and exits the loop
```
uint8_t c = 0;
do {
      Serial.write(c++);
      delay(50);
} while (c); 
```
is what I'd do

**Reply by mosqu-ito** [reply]
Cute, and yes, that's why I used a word, knowing the compiler would (rightly) complain that a byte is ALWAYS <=255.  But in a loop that includes a 50mS delay, efficiency might not be of paramount concern.  I only reluctantly moved to C++ having spent most of my life with Pascal.  I'm all about readability.  But it's lucky for guys like me that there are guys like you in under the hood.   :D

**Comment by mosqu-ito** [comment]
I wonder what happened to OP?  I'll bet the space aliens who were scrambling his comms have taken him up to the mothership for probing.
